pkg/export/table.go

```diff
@@ -0,0 +1,136 @@
+package export
+
+// This formatter differs from github.com/cli/cli/utils.TablePrinter
+// because text/tabwriter.Writer ignores escape sequences like color
+// terminal sequences. This allows users to more easily use --template
+// and relies on text/tabwriter to handle elastic tabstops.
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"text/tabwriter"
+)
+
+type tableFormatter struct {
+	current *table
+	writer  io.Writer
+}
+
+type table struct {
+	tabwriter *tabwriter.Writer
+}
+
+// Initialize a new tabwriter.
+func (f *tableFormatter) Table(args ...interface{}) (string, error) {
+	minwidth, err := intArg(args, 0, 0)
+	if err != nil {
+		return "", fmt.Errorf("failed to initialize table: %v", err)
+	}
+
+	tabwidth, err := intArg(args, 1, 1)
+	if err != nil {
+		return "", fmt.Errorf("failed to initialize table: %v", err)
+	}
+
+	padding, err := intArg(args, 2, 1)
+	if err != nil {
+		return "", fmt.Errorf("failed to initialize table: %v", err)
+	}
+
+	padchar, err := byteArg(args, 3, ' ')
+	if err != nil {
+		return "", fmt.Errorf("failed to initialize table: %v", err)
+	}
+
+	flags, err := uintArg(args, 4, 0)
+	if err != nil {
+		return "", fmt.Errorf("failed to initialize table: %v", err)
+	}
+
+	f.current = &table{
+		// Write directly to the same writer as template execution
+		// since writing to a simple buffer doesn't flush properly.
+		tabwriter: tabwriter.NewWriter(f.writer, minwidth, tabwidth, padding, padchar, flags),
+	}
+	return "", nil
+}
+
+// Write a row of string fields.
+func (f *tableFormatter) Row(fields ...interface{}) (string, error) {
+	if f.current == nil {
+		return "", fmt.Errorf("no current table. use '{{table}}' to start a table")
+	}
+	stringFields := make([]string, len(fields))
+	for i, e := range fields {
+		s, err := jsonScalarToString(e)
+		if err != nil {
+			return "", fmt.Errorf("failed to write row: %v", err)
+		}
+		stringFields[i] = s
+	}
+	row := strings.Join(stringFields, "\t")
```

Comment:
It feels weird that we're serializing individual fields as a tab-delimited string in order to pass it to `tabwriter` that will parse each row and split it into those same fields again, but I guess TabWriter doesn't give us direct access to its implementation of elastic tabstops.

Could we instead use our own TablePrinter as a table renderer mechanism and modify it to ignore ANSI escape sequences when measuring display width of text? https://github.com/muesli/reflow/blob/master/ansi/buffer.go

Reply:
I found `TablePrinter` when writing this and considered using it, but it was comparatively more restrictive. The idea of passing funcs to handle truncation and colorization are novel, but difficult to model in a way that works for templates.

Stepping through more of the tabwriter, it doesn't actually strip vterm sequences in calculations. It simply uses `utf8.RuneCount` on the current field which does include the length of the vterm sequence. I believe the terminal itself (in my case Windows Terminal - I'll set up a Ubuntu VM and test a couple of its terminals but suspect they will do the same since WT is playing catch up) is doing the adjustment of tab stops. I'll dig more into `TablePrinter`'s handling of this, but wonder if it's even worth passing funcs to format or just passing in formatted text.

Reply:
BTW, any particular reason you defined several functions for calculating display widths in pkg/text instead of using `utf8.RuneCount` or `utf8.RuneCountInString`? If the results match and I end up doing some of this refactoring, I wanted to understand if there's any unforeseen reason not to reduce the amount of code to maintain as well.

Reply:
I think I could make `TablePrinter` work if I change `pkg/text.DisplayWidth` to at least skip [vterm color sequences using CSI codes](https://en.wikipedia.org/wiki/ANSI_escape_code#Fe_Escape_sequences) since that's all we should expect currently. The `colorFunc` would just be nil and accept whatever text is added to `AddField`. This way it also uses more of the terminal width by default. It'd also probably obviate the need to expose `pkg/text.Truncate` directly since the table formatting would be automatic, as opposed to forcing the user to assume some typical width to specify.

That said, what I [commented on earlier](https://github.com/cli/cli/pull/3519#issuecomment-828704368) would still be a problem. Perhaps trying to support multiple tables isn't worth the effort. In this case, perhaps the only function we add is `{{row ...}}`. Initializing and rendering a table would happen within `pkg/export.ExecuteTemplate()`. It'd be less authoring in the template so less chance of mistakes.

```
{{range .}}{{row (.number | color "green") .title}}{{end}}
```

Thoughts? I'll start down this path for now.

Reply:
Close, but might have to fix this (ideally) first or duplicate it (less than ideal): https://github.com/muesli/reflow/issues/35

Reply:
Now irrelevant. I've refactored a bit to use TablePrinter.